src/main/java/net/dries007/tfc/common/blocks/devices/CharcoalForgeBlock.java

```diff
@@ -0,0 +1,170 @@
+import java.util.function.BiPredicate;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.enchantment.EnchantmentHelper;
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
+import net.minecraft.particles.ParticleTypes;
+import net.minecraft.state.IntegerProperty;
+import net.minecraft.state.StateContainer;
+import net.minecraft.util.*;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.BlockRayTraceResult;
+import net.minecraft.util.math.shapes.ISelectionContext;
+import net.minecraft.util.math.shapes.VoxelShape;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
+import net.minecraft.world.World;
+import net.minecraft.world.server.ServerWorld;
+import net.minecraftforge.api.distmarker.Dist;
+import net.minecraftforge.api.distmarker.OnlyIn;
+import net.minecraftforge.fml.network.NetworkHooks;
+
+import net.dries007.tfc.common.TFCTags;
+import net.dries007.tfc.common.blocks.CharcoalPileBlock;
+import net.dries007.tfc.common.blocks.ForgeBlockProperties;
+import net.dries007.tfc.common.blocks.TFCBlockStateProperties;
+import net.dries007.tfc.common.tileentity.CharcoalForgeTileEntity;
+import net.dries007.tfc.util.Helpers;
+import net.dries007.tfc.util.MultiBlock;
+
+public class CharcoalForgeBlock extends DeviceBlock
+{
+    public static final IntegerProperty HEAT = TFCBlockStateProperties.HEAT_LEVEL;
+
+    private static final MultiBlock FORGE_MULTIBLOCK;
+
+    static
+    {
+        BiPredicate<IWorld, BlockPos> skyMatcher = IWorld::canSeeSky;
+        BiPredicate<IWorld, BlockPos> isValidSide = Helpers.createTagCheck(TFCTags.Blocks.FORGE_INSULATION);
+        FORGE_MULTIBLOCK = new MultiBlock()
+            // Top block
+            .match(new BlockPos(0, 1, 0), state -> state.isAir() || state.is(TFCTags.Blocks.FORGE_INVISIBLE_WHITELIST))//todo: crucible is also acceptable
+            // Chimney
+            .matchOneOf(new BlockPos(0, 1, 0), new MultiBlock()
+                .match(new BlockPos(0, 0, 0), skyMatcher)
+                .match(new BlockPos(0, 0, 1), skyMatcher)
+                .match(new BlockPos(0, 0, 2), skyMatcher)
+                .match(new BlockPos(0, 0, -1), skyMatcher)
+                .match(new BlockPos(0, 0, -2), skyMatcher)
+                .match(new BlockPos(1, 0, 0), skyMatcher)
+                .match(new BlockPos(2, 0, 0), skyMatcher)
+                .match(new BlockPos(-1, 0, 0), skyMatcher)
+                .match(new BlockPos(-2, 0, 0), skyMatcher)
+            )
+            // Underneath
+            .match(new BlockPos(1, 0, 0), isValidSide)
+            .match(new BlockPos(-1, 0, 0), isValidSide)
+            .match(new BlockPos(0, 0, 1), isValidSide)
+            .match(new BlockPos(0, 0, -1), isValidSide)
+            .match(new BlockPos(0, -1, 0), isValidSide);
```

Comment:
I am *very tempted* to suggest rewriting these like such:
```java
BlockPos origin = BlockPos.ZERO;

// later
.match(origin.east())
.match(origin.down())
.match(origin.east(2))
```
etc.

I think that might look a bit cleaner and be more future proof for if we ever need to edit / understand these values. Thoughts?